newIDE/app/src/MainFrame/Preferences/PreferencesProvider.js

```diff
@@ -343,6 +346,49 @@ export default class PreferencesProvider extends React.Component<Props, State> {
     );
   }
 
+  _getRecentProjectFiles() {
+    return this.state.values.recentProjectFiles;
+  }
+
+  _setRecentProjectFiles(recents: Array<FileMetadata>) {
+    this.setState(
+      state => ({
+        values: {
+          ...state.values,
+          recentProjectFiles: recents,
+        },
+      }),
+      () => this._persistValuesToLocalStorage(this.state)
+    );
+  }
+
+  _insertRecentProjectFile(fileMetadata: FileMetadata) {
+    let recentProjectFiles = this._getRecentProjectFiles();
+    let index = -1;
+    recentProjectFiles.forEach((item, _index) => {
+      if (JSON.stringify(item) === JSON.stringify(fileMetadata)) {
+        index = _index;
+        return;
+      }
+    });
+    if (index === 0) {
+      this._setRecentProjectFiles(recentProjectFiles);
+      return;
+    }
+    if (index === -1 && recentProjectFiles.length === 5)
+      recentProjectFiles.shift();
+    if (index > 0) {
+      const len = recentProjectFiles.length > 5 ? 5 : recentProjectFiles.length;
+      recentProjectFiles = [
+        ...recentProjectFiles.slice(0, index),
+        ...recentProjectFiles.slice(index + 1, len),
+      ];
+    }
+
+    recentProjectFiles = [fileMetadata, ...recentProjectFiles];
+    this._setRecentProjectFiles(recentProjectFiles);
+  }
```

Comment:
Should be decomposable, in 3 simple steps. Add the file at top, remove it from anywhere it was before, ensure the list has a maximum length: 

```suggestion
  _insertRecentProjectFile(newFileMetadata: FileMetadata) {
    const recentProjectFiles = this._getRecentProjectFiles();
    const isNotNewFileMetadata = (fileMetadata) =>
        JSON.stringify(fileMetadata) !== JSON.stringify(newFileMetadata);
    this._setRecentProjectFiles(
        [fileMetadata, ...recentProjectFiles.filter(isNotNewFileMetadata)].slice(
        0,
        MAX_RECENT_FILES
        )
    );
  }
```